### PR TITLE
Update conf only when the container is connectable

### PR DESCRIPTION
**Pull request by dragomirp**
## Issue
[Sometimes](https://github.com/canonical/pgbouncer-k8s-operator/actions/runs/4719644966/jobs/8393434027) there is an unhandled exception while regenerating the patroni configuration due to the charm being unable to read the postgresql version from the workload container

## Solution
Check if the workload container is connectable before regenerating the config

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -232,6 +232,13 @@ def _on_peer_relation_changed(self, event: RelationChangedEvent) -> None:
 
         # Update the list of the cluster members in the replicas to make them know each other.
         # Update the cluster members in this unit (updating patroni configuration).
+        container = self.unit.get_container("postgresql")
+        if not container.can_connect():
+            logger.debug(
+                "Deferring on_peer_relation_changed: Waiting for container to become available"
+            )
+            event.defer()
+            return
```
Comment (dragomirp) [inline src/charm.py:241]: I'm not sure if we should defer here. If the workload is not fully up, the configuration should be generated once pebble is ready. Thoughts?

File: tests/unit/test_patroni.py [hunk 2]
```diff
@@ -166,8 +166,10 @@ def test_render_patroni_yml_file(self, _render_file, _rock_postgresql_version):
         )
         self.assertIn("ssl_key_file: /var/lib/postgresql/data/key.pem", expected_content_with_tls)
 
+    @patch("patroni.stop_after_delay", return_value=stop_after_delay(0))
+    @patch("patroni.wait_fixed", return_value=wait_fixed(0))
```
Comment (dragomirp) [inline tests/unit/test_patroni.py:170]: So we don't wait for tenacity.

File: src/charm.py (hunk 1 above)
Comment (marceloneppel) [reply src/charm.py:241]: I think it's ok to just return here, and leave to pebble ready event to update the config. If, later, the config needs to be updated again, this or some other hooks will be fired again (to call update_config).

**Review by taurus-forever — APPROVED** [review]

**Comment by taurus-forever** [comment]
Nice trace! Tnx!

**Pull request merged**